drivers: led: Provide a fallback implementation of blink API

anobli opened the pull request:
The LED blink API is quite useful but this is not supported by all the LED controller.
This provides a software implementation of blink that use a delayable work to blink a LED.
This could be enabled using Kconfig, as well the maximum number of LED that could blink at the same time using blink fallback. This does not requires any changes to driver to be used.

anobli committed:
drivers: led: Provide a software implementation of blink API

The LED blink API is quite useful but this is not supported by all
the LED controller.
This provides a software implementation of blink that uses a delayable
work to blink a LED.
To use it, a driver just have to allocate and init a struct for holding
blinking led data and to call the software blinking functions.

Signed-off-by: Alexandre Bailon <abailon@baylibre.com>

simonguinot commented on include/zephyr/drivers/led.h:
```diff
@@ -128,6 +128,11 @@ __subsystem struct led_driver_api {
 
+#ifdef CONFIG_LED_BLINK_FALLBACK
+int led_blink_fallback(const struct device *dev, uint32_t led, uint32_t delay_on,
+		       uint32_t delay_off);
```
Suggestion:
```
#ifdef CONFIG_LED_BLINK_SOFTWARE
int led_blink_start(const struct device *dev, uint32_t led,
                       uint32_t delay_on, uint32_t delay_off);
int led_blink_stop(const struct device *dev, uint32_t led);
#else
#define led_blink_start(dev, led) do {} while (0)
#define led_blink_stop(dev, led) do {} while (0)
#endif
```

Or something like that...

anobli replied:
Unless we want to make them "public" API, I prefer to keep them out of led.h.
Since I move some functions to led_core.c, this is not needed here anymore.

simonguinot commented on include/zephyr/drivers/led.h:
```diff
@@ -331,6 +340,13 @@ static inline int z_impl_led_on(const struct device *dev, uint32_t led)
+	if (api->blink == NULL) {
+		led_blink_fallback(dev, led, 0, 0);
+	}
+#endif
```
And then you can just call `led_blink_stop` here.

simonguinot commented on include/zephyr/drivers/led.h:
```diff
@@ -150,7 +155,11 @@ static inline int z_impl_led_blink(const struct device *dev, uint32_t led,
 		return -ENOSYS;
+#endif
 	}
 	return api->blink(dev, led, delay_on, delay_off);
```
Maybe this should be moved into `led-core.c` as well.

simonguinot reviewed (changes requested):
Hi @anobli,

Thanks for this PR but I can't agree with it in its current state. Please see my comments below.

faxe1008 commented:
Random thought: maybe build it in away that allows to build more complex animations (like pulsing, breathing), and let a user provide a step function of sorts to determine the next brightness value.

simonguinot commented:
> Random thought: maybe build it in away that allows to build more complex animations (like pulsing, breathing), and let a user provide a step function of sorts to determine the next brightness value.

It should be discussed but I think that more complex LED effects (run-light, breathing, rainbow, etc) and other things (such as synchronization between devices, color correction, etc) should go in a user library, which is what we are missing the most right now. Blinking is a special case to me. It is cheap and commonly used by simple applications. For theses reasons I think it makes sense to provide a software implementation in Zephyr kernel.

thedjnK reviewed (dismissed):
the whole point in a driver is that it controls hardware, this is doing a complete software thing, it has no place in drivers, if this is wanted then it should be a software component i.e. go in subsys

fabiobaltieri commented on drivers/led/blink_fallback.c:
```diff
@@ -0,0 +1,130 @@
+
+static void led_blink_fallback_off(struct k_work *work);
+
+static K_MUTEX_DEFINE(led_blink_fallback_mutex);
```
use a semaphore, lighter implementation, this does not need the extra stuff that comes with mutextes. https://youtu.be/p8OgqVQxklo?si=z8xJ5mp_Qi86zO9V&t=1305

anobli replied:
I use static allocation now so this should not be needed anymore.

fabiobaltieri commented on drivers/led/blink_fallback.c:
```diff
@@ -0,0 +1,130 @@
+
+	const struct led_driver_api *api = (const struct led_driver_api *)data->dev->api;
+
+	if (api->on) {
+		api->on(data->dev, data->led);
+	} else {
+		api->set_brightness(data->dev, data->led, LED_BRIGTHNESS_MAX);
+	}
```
just call `led_on`, it'll fall back to the right thing for you

fabiobaltieri commented on drivers/led/blink_fallback.c:
```diff
@@ -0,0 +1,130 @@
+
+	const struct led_driver_api *api = (const struct led_driver_api *)data->dev->api;
+
+	if (api->off) {
+		api->off(data->dev, data->led);
+	} else {
+		api->set_brightness(data->dev, data->led, 0);
+	}
```
same, `led_off`

anobli committed:
drivers: led: gpio: Add support of software blink

This updates the led_gpio driver to use software blinking.

Signed-off-by: Alexandre Bailon <abailon@baylibre.com>

anobli committed:
drivers: led: Create led_core.c

In order to develop a sotfware implementation for led blinking,
prepare the led API.
This move z_impl_led_on and z_impl_led_off code to led_core.c
so we could use and update them for blinking a led.

Signed-off-by: Alexandre Bailon <abailon@baylibre.com>

anobli committed:
drivers: led: Add a new callback to get software blinking data

Blinking data are expected to be attached to led device.
In order to have a as much as possible generic implementation,
we need a way to retreive the data from the drivers.
This adds a new callback that will be used in that purpose.

Signed-off-by: Alexandre Bailon <abailon@baylibre.com>

anobli commented on drivers/led/blink_fallback.c:
```diff
@@ -0,0 +1,130 @@
+	struct k_work_delayable work;
+	uint32_t delay_on;
+	uint32_t delay_off;
+};
```
I have added a new struct attached to the LED device. I tried to do it in a way that request less changes as possible in driver. I kept the same content in the struct. I think I can probably improve it a little bit by removing the led index and use pointer arithmetic to retrieve it.

anobli commented on drivers/led/blink_fallback.c:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	/* Stop blinking, more likely when we use led_on or led_off */
+	if (delay_on == 0 && delay_off == 0) {
```
Done! And for the specific case where both delay are 0, I use the policy as Linux: 1Hz blinking. I hope this makes sense.

anobli commented:
> the whole point in a driver is that it controls hardware, this is doing a complete software thing, it has no place in drivers, if this is wanted then it should be a software component i.e. go in subsys

I am confuse here. Are you talking about the suggestion from @faxe1008 or about doing software blinking in drivers?
If this about doing software blinking in drivers, I disagree. Although there is no hardware support, we are implementing driver API so I think it must be located here.

anobli commented:
I did a lot of change since the previous version. I hope this is going in the direction even if some improvement are doable.
Software blinking is more complex than I was thinking!
I have experimented different implementation with different pro and cons, and I have not found a solution that fully satisfied me ...
This new version of the PR should have good tradeoff between complexity, memory footprint and the number of change required to enable it in a driver.
I have implemented only the software blinking for the led_gpio driver. I could update the other drivers later once this one get approved.
